lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+
+    /// Rescale the input unitary matrix, `u`, to be special unitary.
+    /// Extract a phase factor, `phase`, so that `det(inv_phase * unitary) = 1`
+    auto det = (matrix[0] * matrix[3]) - (matrix[1] * matrix[2]);
+    phase = 0.5 * std::atan2(det.imag(), det.real());
+    auto inv_phase = std::exp(-1i * phase);
+
+    std::array<std::complex<double>, 4> special_unitary;
+    std::transform(matrix.begin(), matrix.end(), special_unitary.begin(),
+                   [&](auto element) { return element * inv_phase; });
+
+    auto abs_00 = std::abs(special_unitary[0]);
+    auto abs_01 = std::abs(special_unitary[1]);
+
+    if (abs_00 >= abs_01)
+      angles.beta = 2 * std::acos(abs_00);
+    else
+      angles.beta = 2 * std::asin(abs_01);
+
+    auto half_sum_alpha_gamma =
+        std::atan2(special_unitary[3].imag(), special_unitary[3].real());
+    auto half_diff_alpha_gamma =
+        std::atan2(special_unitary[2].imag(), special_unitary[2].real());
+
+    angles.alpha = half_sum_alpha_gamma + half_diff_alpha_gamma;
+    angles.gamma = half_sum_alpha_gamma - half_diff_alpha_gamma;
+  }
+
+  BasisZYZ(std::vector<std::complex<double>> vec) {
+    std::move(vec.begin(), vec.end(), matrix.begin());
+  }
+};
+
+class CustomUnitaryPattern
+    : public OpRewritePattern<quake::CustomUnitarySymbolOp> {
+
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::CustomUnitarySymbolOp customOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto parentModule = customOp->getParentOfType<ModuleOp>();
+
+    Location loc = customOp->getLoc();
+    auto targets = customOp.getTargets();
+    auto controls = customOp.getControls();
+
+    /// Get the global constant holding the concrete matrix corresponding to
+    /// this custom operation invocation
+    StringRef generatorName = customOp.getGenerator().getRootReference();
+    auto globalOp =
+        parentModule.lookupSymbol<cudaq::cc::GlobalOp>(generatorName);
+
+    /// The decomposed sequence of quantum operations are in a function
+    auto pair = generatorName.split(".rodata");
+    std::string funcName = pair.first.str() + ".kernel" + pair.second.str();
+
+    /// If the replacement function doesn't exist, create it here
+    if (!parentModule.lookupSymbol<LLVM::LLVMFuncOp>(funcName)) {
+
+      auto unitary = cudaq::readGlobalConstantArray(globalOp);
+
+      /// TODO: Expand the logic to decompose upto 4-qubit operations
+      if (unitary.size() != 4)
+        return customOp.emitError(
+            "Decomposition of only single qubit custom operations supported.");
+
+      /// Controls are handled via apply specialization
+      auto funcTy =
+          FunctionType::get(parentModule.getContext(), targets.getTypes(), {});
+      auto insPt = rewriter.saveInsertionPoint();
+      rewriter.setInsertionPointToStart(parentModule.getBody());
+      auto func = rewriter.create<func::FuncOp>(parentModule->getLoc(),
```

Comment:
This looks like a lurking multithreading bug. Since this pass is specified as a `func::FuncOp` pass, we shouldn't be modifying the Module here. Doing so is a silent race condition which will cause mysterious random crashes.

Reply:
Thanks! Addressed in 3e71dee865f261ba1ba9379e172eb4b68ccab2d7 by changing to a module pass.